unstructured/partition/docx.py

```diff
@@ -134,305 +170,625 @@ def partition_docx(
+
+    @lazyproperty
+    def _blocks_in_and_above_section_xpath(self) -> str:
+        """XPath expr for ps and tbls in context of a sectPr and all prior sectPrs."""
+        # -- "p_sect" is a section with sectPr located at w:p/w:pPr/w:sectPr. "body_sect" is a
+        # -- section with sectPr located at w:body/w:sectPr. The last section in the document is a
+        # -- "body_sect". All others are of the "p_sect" variety. "term" means "terminal", like
+        # -- the last p or tbl in the section. "pred" means "predecessor", like a preceding p or
+        # -- tbl in the section.
+
+        # -- the terminal block in a p-based sect is the p the sectPr appears in --
+        p_sect_term_block = "./parent::w:pPr/parent::w:p"
+        # -- the terminus of a body-based sect is the sectPr itself (not a block) --
+        body_sect_term = "self::w:sectPr[parent::w:body]"
+        # -- all the ps and tbls preceding (but not including) the context node --
+        pred_ps_and_tbls = "preceding-sibling::*[self::w:p | self::w:tbl]"
+
+        # -- p_sect_term_block and body_sect_term(inus) are mutually exclusive. So the result is
+        # -- either the union of nodes found by the first two selectors or the nodes found by the
+        # -- last selector, never both.
+        return (
+            # -- include the p containing a sectPr --
+            f"{p_sect_term_block}"
+            # -- along with all the blocks that precede it --
+            f" | {p_sect_term_block}/{pred_ps_and_tbls}"
+            # -- or all the preceding blocks if sectPr is body-based (last sectPr) --
+            f" | {body_sect_term}/{pred_ps_and_tbls}"
+        )
 
-                    if _type == "header":
-                        headers.append(Header(text=text, metadata=metadata))
-                    elif _type == "footer":
-                        footers.append(Footer(text=text, metadata=metadata))
+    def _count_of_blocks_in_and_above_section(self, sectPr: CT_SectPr) -> int:
+        """All ps and tbls in section defined by `sectPr` and all prior sections."""
+        if self._compiled_count_xpath is None:
+            self._compiled_count_xpath = etree.XPath(
+                f"count({self._blocks_in_and_above_section_xpath})",
+                namespaces=nsmap,
+                regexp=False,
+            )
+        xpath = self._compiled_count_xpath
+        # -- numeric XPath results are always float, so need an int() conversion --
+        return int(cast(float, xpath(sectPr)))
+
+    @lazyproperty
+    def _sectPrs(self) -> Sequence[CT_SectPr]:
+        """All w:sectPr elements in document, in document-order."""
+        return self._sectPr.xpath(
+            "/w:document/w:body/w:p/w:pPr/w:sectPr | /w:document/w:body/w:sectPr"
+        )
 
-        headers_and_footers.append((headers, footers))
 
-    return headers_and_footers
+# == monkey-patch docx.text.Paragraph.runs ===========================================
```

Comment:
Hi Klaijan, that "divider" comment is meant to visually segregate the `_get_paragraph_runs()` function and the `Paragraph.runs = property(...` statement below it because those actually modify the `python-docx` library code and need to move upstream. So this comment line and the one on the last line of the file are a "matched pair" indicating the boudaries of that block.

I placed it at the bottom for two reasons, one because it is not referenced by the docx-partitioning code in the module and two because it will be going away soon when we move it upstream.